From 1e958818ffbf3e5d69835384bb7e84e3044c0312 Mon Sep 17 00:00:00 2001
From: Richard Yu <richard.yu@mitrastar.com.tw>
Date: Tue, 13 May 2025 20:25:59 +0800
Subject: [PATCH 7/7] EventLoopImpl: refactor run() to prioritize handling
 EINTR
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Refactored EventLoopImpl::run() to prioritize handling EINTR during
epoll_wait(). Now, when epoll_wait() is interrupted by a signal,
run() exits immediately, allowing the main loop to invoke the
appropriate signal handler before calling run() again. This improves
the application's responsiveness to signals like SIGTERM, ensuring
they are handled promptly.

Previously, with the epoll_wait() do-while retry loop, the signal
handler would not execute until the next RX event arrived or the
blocking epoll_wait() timed out. This could result in unpredictable
and undesirable delays — in some cases, where epoll_wait() blocked
indefinitely, it took over 4 seconds for the next RX event to arrive,
leading to a 4+ second delay in responding to SIGTERM.

https://prplfoundationcloud.atlassian.net/browse/PPM-3204

Signed-off-by: Richard Yu <richard.yu@mitrastar.com.tw>
Co-authored-by: Lu Dai <lu.dai@mind.be>
Signed-off-by: Lu Dai <lu.dai@mind.be>
---
 .../bcl/source/beerocks_event_loop_impl.cpp        | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)
 mode change 100644 => 100755 common/beerocks/bcl/source/beerocks_event_loop_impl.cpp

diff --git a/common/beerocks/bcl/source/beerocks_event_loop_impl.cpp b/common/beerocks/bcl/source/beerocks_event_loop_impl.cpp
old mode 100644
new mode 100755
index 4f5beffe3..b5b614aaf
--- a/common/beerocks/bcl/source/beerocks_event_loop_impl.cpp
+++ b/common/beerocks/bcl/source/beerocks_event_loop_impl.cpp
@@ -169,15 +169,15 @@ int EventLoopImpl::run()
         (m_timeout > std::chrono::milliseconds::zero()) ? static_cast<int>(m_timeout.count()) : -1;
 
     // Poll the file descriptors
-    // Retry if the call was interrupted by a signal handler before either (1) any of the
-    // requested events occurred or (2) the timeout expired
-    int num_events;
-    do {
-        num_events = epoll_wait(m_epoll_fd, events, events_size, timeout_millis);
-    } while ((num_events < 0) && (EINTR == errno));
+    int num_events = epoll_wait(m_epoll_fd, events, events_size, timeout_millis);
+    int err        = errno; // Save errno before any logging
 
     if (num_events == -1) {
-        LOG(ERROR) << "Error during epoll_wait: " << strerror(errno);
+        if (EINTR == err) {
+            LOG(INFO) << strerror(err) << " during epoll_wait";
+            return 0;
+        }
+        LOG(ERROR) << "Error during epoll_wait: " << strerror(err);
         return -1;
     }
 
-- 
2.34.1

